wallet-new/src/Cardano/Wallet/API/V1/LegacyHandlers/Transactions.hs

```diff
@@ -67,6 +67,25 @@ newTransaction submitTx Payment {..} = do
     cTx <- V0.newPaymentBatch submitTx spendingPw batchPayment
     single <$> migrate cTx
 
+newTransactionFeesIncluded
+    :: forall ctx m . (V0.MonadWalletTxFull ctx m)
+    => (TxAux -> m Bool) -> Payment -> m (WalletResponse Transaction)
+newTransactionFeesIncluded submitTx pmt = do
+    EstimatedFees (V1 coin) <- wrData <$> estimateFees pmt
+    let outputs = length (pmtDestinations pmt)
+        coinPerOutput = Core.getCoin coin `div` fromIntegral outputs
+        newPayment = pmt
+            { pmtDestinations =
+                map subtractAmount (pmtDestinations pmt)
+            }
+        subtractAmount pmtDistribution = pmtDistribution
+            { pdAmount = V1 $
+                (unV1 (pdAmount pmtDistribution))
+                    `Core.unsafeSubCoin`
+                        Core.mkCoin coinPerOutput
+            }
+
+    newTransaction submitTx newPayment
```

Comment:
So, this is a somewhat dumb first pass implementation. Basically, we calculate the estimated fee for the payment, and then subtract that amount from the `pmtDestinations` evenly (eg if there is one destination, then we subtract the entire fee from that destination, if there are two destinations, we subtract half of the fee from each destination, etc).

I'm not convinced it's a great strategy. But I think it *should* work. Some issues:

- A `PaymentDistribution` might be reduced to `0` coins. in this case, we probably want to throw an error, because that's very likely not what they want.
- The user might want to specify a single payment from which the fee comes out of. Embedding the `feesIncluded` into the `PaymentDistribution` object itself would allow that level of control, but would make the logic much trickier.
- We might want to specify different strategies for distributing the fee among the distributions. Right now it's 100% evenly split, though we might want to say "distribute the fee proportionally to each output depending on how large the output is." This isn't that much more difficult to implement logic-wise.

That we have multiple possible strategies suggests that `includeFees` query param should accept the strategy we want to use for distributing included fees. A few fixed strategies will be simpler and easier than embedding the `includedFees` into the payment type itself.

Reply:
Although it seems more acceptable in the crypto world because of rather _unpredictable fees_, I am bit worried about divisions and rounding issues here.
I'd also find more logical to distribute fees according to the amounts sent to each destination as you suggest in your second point but I am pretty sure that we will find people on both sides. What about leaving this as a choice for the user? 

e.g.
```hs
data FeeHeuristic = Exclusive | InclusiveEvenly | InclusiveProportionally 
```

Then, we can build more heuristic over time if needed, that could also rely on fixed strategies ?

Reply:
mmm, division/rounding errors are a good point.

Another potential strategy is to entirely drain the "change output" before subtracting anything from the other outputs. If the change output is entirely drained (eg reduced to 0), then we can remove that output from the transaction and get a lower fee.

We'll also need to be careful of any non-change outputs that are reduced to 0 for this.

The more I'm thinking on this, the more I feel like this shouldn't *actually* submit the transaction, but should instead render a prepared transaction that can later be submitted. Eg: 

> Client: I give you this `Payment` and I want the fees to come out of the `PaymentDistributions` according to this `FeeStrategy`.
> Server: OK, here's the transaction that would result -- these outputs would be reduced by this much. If that is OK, then submit this `Payment` to the ordinary transaction endpoint and it'll do what you want.

Reply:
I like the confirmation idea, though it's an extra step. This _simple_ tasks sounds more like a story to me and by that I mean, it should go over a proper design and discussion phase, testing, and implementation for there are many strategies we could take here. An informal specification on our approach that can be reviewed / documented would be nice. 

What do you think?

Reply:
I've decided that I want to only allow proportional distribution of fee among outputs; the other strategies that I can imagine are Bad (for reasons that will become evident when I push my latest commit with a ton of docs :sweat_smile: )

Reply:
@parsonsmatt @KtorZ Sorry, late to the party, had a looong hacking week 😉 

> We'll also need to be careful of any non-change outputs that are reduced to 0 for this.

I would exercise caution (I know this is a sketch, but still nice to know the fuller picture) to change anything which was originally scheduled in a transaction by the input selection mechanism, as this could have impacts on the anonymity of users. In particular, I would defer any change in the inner logic (or perhaps this ticket altogether) up until @edsko formalises the input selection policies as part of the new data layer (see: CBR-226, CBR-241, CBR-251). What I'm concerned about here is that we are stepping on each other toes -- looks like there are lots of valuable ideas from an API perspective in this PR and I think would be nice to see if they can retrofit into Edsko's policies, but at the same time we should make sure what we offer at the API level doesn't contradict directly the new data layer internal implementation. Am I making any sense? 😉

Reply:
@adinapoli-iohk I got a similar feeling to be honest. This task seems a bit more complicated than it seems, with quite some important choices to make when it comes to selecting inputs. That's definitely not trivial. 
Should we put this task on hold for now and move it under CBR-227 ?

Reply:
> This task seems a bit more complicated than it seems, with quite some important choices to make when it comes to selecting inputs. That's definitely not trivial.

Yea, unfortunately this _seems_ easy when it comes to the API side of it, but as it touches a very delicate part of the wallet, I have rapidly realised this was not going to be trivial. Matt will kill me as it's the second time to does work in this area and he has to pause his hacking 😬 

> Should we put this task on hold for now and move it under CBR-227 ?

I don't see nothing wrong with putting this in a user story (if you folks have one) related to "V2 API" or "API future improvements" and have it depends from CBR-227.

Different matter is if we think we cannot deliver this endpoint without having an extra requirement to be added in @edsko 's input selection policy mechanism, at which point it might make sense to consider the backend side of it as part of CBR-227, and the fronted side as per above.

Reply:
> I don't see nothing wrong with putting this in a user story (if you folks have one) related to "V2 API" or "API future improvements" and have it depends from CBR-227.

Fun fact: this task is currently part of a story about "wrapping up remaining work for **V1**". I guess we were too optimistic about it. That's clearly not something that is part of the _minimal_ V1.

Reply:
This logic is operating at a slightly higher level than input selection. Right now we're just taking a `Payment`, calculating the estimated fee, then reducing the amounts in the `Payment` record. It's possible that this new `Payment` gets a different fee, so I suppose we'd need to iterate that process until it stabilized.

We would not want to take anything from change outputs -- it turns out, that would defeat the point entirely, as the entire intent is to have the fee come from the amount we are *sending* and not the amount we *own*.

I don't feel like it's terribly complicated, just some fiddly bits to minimize the amount of wasted coin when dealing with the rounding. Moving it to a new user story would help get the V1 thing out the door though :)